chainerx_cc/chainerx/routines/normalization.cc

```diff
@@ -117,43 +273,70 @@ Array BatchNorm(
-
     const Array& gamma_reshaped = result.gamma;
     const Array& beta_reshaped = result.beta;
+    const Array& mean_reshaped = result.mean;
+    const Array& var_reshaped = result.var;
+    const Axes& sorted_axis = result.sorted_axis;
+
+    Device& device = x.device();
+
+    // Compute forward.
+    Array out = EmptyLike(x, device);
+    std::shared_ptr<BatchNormState> state{};
+    {
+        NoBackpropModeScope scope{};
+        std::tuple<Array, std::shared_ptr<BatchNormState>> batch_norm_result = device.backend().CallOp<BatchNormOp>(
+                x.AsGradStopped(),
+                gamma_reshaped.AsGradStopped(),
+                beta_reshaped.AsGradStopped(),
+                mean_reshaped,
+                var_reshaped,
+                eps,
+                decay,
+                sorted_axis,
+                true,
+                out);
+        state = std::get<1>(batch_norm_result);
+    }
+    CHAINERX_ASSERT(state != nullptr);
 
-    Array out = fb->Forward(x.AsGradStopped(), gamma_reshaped.AsGradStopped(), beta_reshaped.AsGradStopped());
-    CHAINERX_ASSERT(out.dtype() == x.dtype());
     internal::MakeViewForForwardBackwardOutput(out);
 
     BackwardBuilder bb{"batch_norm", {x, gamma_reshaped, beta_reshaped}, {out}};
     if (BackwardBuilder::Target bt = bb.CreateTarget({0, 1, 2})) {
-        bt.Define([fb = std::move(fb), x_tok = bb.RetainInput(0), gamma_tok = bb.RetainInput(1), eps, sorted_axis = result.sorted_axis](
-                          BackwardContext& bctx) {
+        bt.Define([state = std::move(state),
+                   x_tok = bb.RetainInput(0),
+                   gamma_tok = bb.RetainInput(1),
+                   eps,
+                   sorted_axis,
+                   beta_shape = beta_reshaped.shape(),
+                   beta_dtype = beta_reshaped.dtype()](BackwardContext& bctx) mutable {
             const Array& gout = *bctx.output_grad();
+            const Array& x = bctx.GetRetainedInput(x_tok);
+            const Array& gamma_reshaped = bctx.GetRetainedInput(gamma_tok);
 
-            Array gx{};
-            Array ggamma{};
-            Array gbeta{};
-            {
-                std::array<Array, 3> ginputs = fb->Backward(gout.AsGradStopped());
-                internal::MakeViewForForwardBackwardOutput(ginputs);
+            Device& device = x.device();
 
-                gx = std::move(ginputs[0]);
-                ggamma = std::move(ginputs[1]);
-                gbeta = std::move(ginputs[2]);
+            // Compute backward.
+            Array gx = EmptyLike(x, device);
+            Array ggamma = EmptyLike(gamma_reshaped, device);
+            Array gbeta = Empty(beta_shape, beta_dtype, device);
+            {
+                NoBackpropModeScope scope{};
+                device.backend().CallOp<BatchNormGradOp>(x, gamma_reshaped, gout, eps, sorted_axis, state, gx, ggamma, gbeta);
```

Comment:
Let's skip allocating output arrays at the caller, as discussed.

Reply:
Fixed. But I have to say I don't think this necessarily looks cleaner.

Reply:
You should be able to use `std::tie` to assign three variables at once.

Reply:
Oops, `std::tie` does not work for `std::array`?

Reply:
Changed to tuple.